// material-ui-filter/lib/components/OperatorField.js
'use strict';

exports.__esModule = true;
exports.OperatorField = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Input = require('@material-ui/core/Input');

var _Input2 = _interopRequireDefault(_Input);

var _styles = require('@material-ui/core/styles');

var _Toolbar = require('@material-ui/core/Toolbar');

var _Toolbar2 = _interopRequireDefault(_Toolbar);

var _Tooltip = require('@material-ui/core/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _IconButton = require('@material-ui/core/IconButton');

var _IconButton2 = _interopRequireDefault(_IconButton);

var _Icon = require('@material-ui/core/Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _selectors = require('../store/selectors');

var filterSelectors = _interopRequireWildcard(_selectors);

var _muishift = require('muishift');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = {};

var OperatorField = exports.OperatorField = function (_Component) {
  _inherits(OperatorField, _Component);

  function OperatorField() {
    var _temp, _this, _ret;

    _classCallCheck(this, OperatorField);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _Component.call.apply(_Component, [this].concat(args))), _this), _this.getFieldType = function (currentField) {
      var fields = _this.props.fields;


      var fieldType = '';

      fields.map(function (field) {
        if (field.name === currentField.value) {
          fieldType = field.type;
        }
        return field;
      });

      return fieldType;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  OperatorField.prototype.render = function render() {
    var _props = this.props,
        queryIndex = _props.queryIndex,
        currentField = _props.currentField,
        query = _props.query,
        fields = _props.fields,
        operators = _props.operators,
        handleQueryChange = _props.handleQueryChange,
        formatMessage = _props.formatMessage,
        classes = _props.classes,
        onClick = _props.onClick;

    var _filterSelectors$sele = filterSelectors.selectQueryProps(query),
        operator = _filterSelectors$sele.operator,
        isCaseSensitive = _filterSelectors$sele.isCaseSensitive;

    if (queryIndex == null || currentField == null || query == null || handleQueryChange == null || fields == null) {
      return _react2.default.createElement('div', null);
    }

    var fieldType = this.getFieldType(currentField);
    var divFields = [];

    operators.map(function (operator) {
      if (operator.type === fieldType || operator.type === 'string' && fieldType === undefined) {
        operator.operators.map(function (op) {
          return divFields.push({
            value: op.value,
            label: op.label
          });
        });
      }
      return divFields;
    });

    return _react2.default.createElement(
      _Toolbar2.default,
      null,
      _react2.default.createElement(
        'div',
        { style: { maxWidth: 160 } },
        _react2.default.createElement(_muishift.SelectField, {
          input: { value: operator },
          onChange: function onChange(val) {
            handleQueryChange(queryIndex, 'operator', val);
          },
          items: divFields,
          itemToString: function itemToString(item) {
            return item ? item.label : '';
          },
          id: 'react-select-single',
          inputProps: {
            fullWidth: true,
            placeholder: formatMessage ? formatMessage({ id: 'hint_autocomplete' }) : 'Select operator'
          }
        })
      ),
      _react2.default.createElement(
        _Tooltip2.default,
        {
          id: 'tooltip-bottom-start',
          title: formatMessage ? formatMessage({ id: isCaseSensitive ? 'disable_case_sensitivity' : 'enable_case_sensitivity' }) : '',
          placement: 'bottom-end' },
        _react2.default.createElement(
          _IconButton2.default,
          { onClick: onClick, 'aria-label': 'Delete', color: 'secondary' },
          _react2.default.createElement(
            _Icon2.default,
            null,
            'delete'
          )
        )
      )
    );
  };

  return OperatorField;
}(_react.Component);

exports.default = (0, _styles.withStyles)(styles)(OperatorField);
